File: src/co/nums/intellij/aem/htl/highlighter/HtlSyntaxHighlighter.java
package co.nums.intellij.aem.htl.highlighter;

import co.nums.intellij.aem.htl.parsing.HtlLexerAdapter;
import co.nums.intellij.aem.htl.psi.HtlTokenTypes;
import com.intellij.lexer.Lexer;
import com.intellij.openapi.editor.colors.TextAttributesKey;
import com.intellij.openapi.fileTypes.SyntaxHighlighterBase;
import com.intellij.psi.tree.IElementType;
import org.jetbrains.annotations.NotNull;

import java.util.HashMap;
import java.util.Map;

import static com.intellij.openapi.editor.DefaultLanguageHighlighterColors.KEYWORD;
import static com.intellij.openapi.editor.DefaultLanguageHighlighterColors.NUMBER;
import static com.intellij.openapi.editor.DefaultLanguageHighlighterColors.STRING;

public class HtlSyntaxHighlighter extends SyntaxHighlighterBase {

	private static final Map<IElementType, TextAttributesKey> ATTRIBUTES = new HashMap<>();

	static {
		fillMap(ATTRIBUTES, KEYWORD, HtlTokenTypes.BOOLEAN_TRUE, HtlTokenTypes.BOOLEAN_FALSE);
		fillMap(ATTRIBUTES, KEYWORD, HtlTokenTypes.EXPR_START, HtlTokenTypes.EXPR_END, HtlTokenTypes.OPTIONS_SEPARATOR);
		fillMap(ATTRIBUTES, STRING, HtlTokenTypes.SINGLE_QUOTED_STRING, HtlTokenTypes.DOUBLE_QUOTED_STRING);
		fillMap(ATTRIBUTES, NUMBER, HtlTokenTypes.INTEGER_NUMBER, HtlTokenTypes.FLOAT_NUMBER);
	}

	@NotNull
	@Override
	public Lexer getHighlightingLexer() {
		return new HtlLexerAdapter();
	}

	@NotNull
	@Override
	public TextAttributesKey[] getTokenHighlights(IElementType tokenType) {
		return pack(ATTRIBUTES.get(tokenType));
	}

}

File: src/main/kotlin/co/nums/intellij/aem/htl/highlighter/HtlSyntaxHighlighter.kt
package co.nums.intellij.aem.htl.highlighter

import co.nums.intellij.aem.htl.lexer.HtlLexerAdapter
import co.nums.intellij.aem.htl.psi.HtlTypes
import com.intellij.openapi.editor.DefaultLanguageHighlighterColors.KEYWORD
import com.intellij.openapi.editor.DefaultLanguageHighlighterColors.NUMBER
import com.intellij.openapi.editor.DefaultLanguageHighlighterColors.STRING
import com.intellij.openapi.editor.XmlHighlighterColors.HTML_COMMENT
import com.intellij.openapi.editor.colors.TextAttributesKey
import com.intellij.openapi.fileTypes.SyntaxHighlighterBase
import com.intellij.psi.tree.IElementType

object HtlSyntaxHighlighter : SyntaxHighlighterBase() {

    private val ATTRIBUTES = mapOf(
            HtlTypes.EXPR_START to KEYWORD,
            HtlTypes.EXPR_END to KEYWORD,
            HtlTypes.OPTIONS_SEPARATOR to KEYWORD,
            HtlTypes.SINGLE_QUOTE to STRING,
            HtlTypes.DOUBLE_QUOTE to STRING,
            HtlTypes.STRING_CONTENT to STRING,
            HtlTypes.INTEGER_NUMBER to NUMBER,
            HtlTypes.FLOAT_NUMBER to NUMBER,
            HtlTypes.BOOLEAN_LITERAL to KEYWORD,
            HtlTypes.COMMENT_START to HTML_COMMENT,
            HtlTypes.COMMENT_CONTENT to HTML_COMMENT,
            HtlTypes.COMMENT_END to HTML_COMMENT
    )

    override fun getHighlightingLexer() = HtlLexerAdapter()

    override fun getTokenHighlights(tokenType: IElementType): Array<out TextAttributesKey> =
            SyntaxHighlighterBase.pack(ATTRIBUTES[tokenType])

}
